File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = set()
        memo = {}
        for num in nums:
            if num in memo:
                memo[num] += 1
            else:
                memo[num] = 1
        
        if 0 in memo and memo[0] >= 3:
            ans.add((0,0,0))
        
        nums = list(set(nums))
        n = len(nums)

        for i in range(n):
            for j in range(i+1, n):
                if -(nums[i]+nums[j]) in memo:
                    if -(nums[i]+nums[j]) != nums[i] and -(nums[i]+nums[j]) != nums[j]:
                        ans.add(tuple(sorted([-(nums[i]+nums[j]), nums[i], nums[j]])))
                    
                    if -(nums[i]+nums[j]) == nums[i] or -(nums[i]+nums[j]) == nums[j]:
                        if memo[-(nums[i]+nums[j])] > 1:
                            ans.add(tuple(sorted([-(nums[i]+nums[j]), nums[i], nums[j]])))
        
        
        return ans

File: 0015-3sum/0015-3sum.java
import java.util.*;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> answer = new ArrayList<>();
        
        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            if (nums[i] > 0) {
                break;
            }

            int lt = i + 1;
            int rt = nums.length - 1;

            while (lt < rt) {
                int sum = nums[i] + nums[lt] + nums[rt];

                if (sum < 0) {
                    lt++;
                } else if (sum > 0) {
                    rt--;
                } else {
                    answer.add(Arrays.asList(nums[i], nums[lt], nums[rt]));
                    lt++;
                    rt--;

                    while (lt < rt && nums[lt] == nums[lt - 1]) {
                        lt++;
                    }

                    while (lt < rt && nums[rt] == nums[rt + 1]) {
                        rt--;
                    }
                }
            }
        }

        return answer;
    }
}